### PR TITLE
[stdlib] Deprecate MutableCollection._withUnsafeMutableBufferPointerIfSupported

### PR DESCRIPTION
**Pull request by lorentey**
In Swift 5.0, [SE-0237] introduced the public `MutableCollection.withContiguousMutableStorageIfAvailable` method. It’s time we migrated off the old, underscored variant and deprecated it.

The default `MutableCollection.sort` and `.partition(by:)` implementations are currently calling this hidden method rather than the documented interface, preventing custom Collection implementations from achieving good performance, even if they have contiguous storage.

[SE-0237]: https://github.com/apple/swift-evolution/blob/master/proposals/0237-contiguous-collection.md

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
Oh, nice, we have test coverage! ❤️ 

```
[ RUN      ] ArraySliceWithNonZeroStartIndex_MutableRandomAccessCollectionRef.ArraySliceWithNonZeroStartIndex.ArraySlice<LifetimeTracked>.Type.partition/DispatchesThrough_withUnsafeMutableBufferPointerIfSupported
stdout>>> check failed at /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-main/swift/stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift, line 968
stdout>>> expected: 1 (of type Swift.Int)
stdout>>> actual: 0 (of type Swift.Int)
stdout>>> check failed at /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-main/swift/stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift, line 970
stdout>>> expected: 1 (of type Swift.Int)
stdout>>> actual: 0 (of type Swift.Int)
[     FAIL ] ArraySliceWithNonZeroStartIndex_MutableRandomAccessCollectionRef.ArraySliceWithNonZeroStartIndex.ArraySlice<LifetimeTracked>.Type.partition/DispatchesThrough_withUnsafeMutableBufferPointerIfSupported
```

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
```
16:33:26 
/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-main/swift/stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift:517:1: 
error: expressions are not allowed at the top level
```

\<shameful facepalm\>

**Comment by lorentey** [comment]
@swift-ci test

File: stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift [hunk 1]
```diff
@@ -965,11 +971,23 @@ self.test("\(testNamePrefix).partition/DispatchesThrough_withUnsafeMutableBuffer
     return !(extractValue(val).value < extractValue(first!).value)
   })
 
-  expectEqual(
-    1, lc.log._withUnsafeMutableBufferPointerIfSupported[type(of: lc)])
+  let actualWUMBPIF = lc.log._withUnsafeMutableBufferPointerIfSupported[type(of: lc)]
+  let actualWCMSIA =  lc.log.withContiguousMutableStorageIfAvailable[type(of: lc)]
+
+  let actualWUMBPIFNonNil = lc.log._withUnsafeMutableBufferPointerIfSupportedNonNilReturns[type(of: lc)]
+  let actualWUMBPIFNonNil =  lc.log.withContiguousMutableStorageIfAvailableNonNilReturns[type(of: lc)]
```
Comment (natecook1000) [inline stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift:978]:
```suggestion
  let actualWCMSIAIFNonNil =  lc.log.withContiguousMutableStorageIfAvailableNonNilReturns[type(of: lc)]
```

File: stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift [hunk 2]
```diff
@@ -644,13 +647,16 @@ self.test("\(testNamePrefix).sorted/DispatchesThrough_withUnsafeMutableBufferPoi
   let result = lc.sorted { extractValue($0).value < extractValue($1).value }
   let extractedResult = result.map(extractValue)
 
+  let actualWUMBPIF = lc.log._withUnsafeMutableBufferPointerIfSupported[type(of: lc)]
+  let actualWCMSIA =  lc.log.withContiguousMutableStorageIfAvailable[type(of: lc)]
+
+  let actualWUMBPIFNonNil = lc.log._withUnsafeMutableBufferPointerIfSupportedNonNilReturns[type(of: lc)]
+  let actualWUMBPIFNonNil =  lc.log.withContiguousMutableStorageIfAvailableNonNilReturns[type(of: lc)]
```
Comment (natecook1000) [inline stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift:654]:
```suggestion
  let actualWCMSIAIFNonNil =  lc.log.withContiguousMutableStorageIfAvailableNonNilReturns[type(of: lc)]
```

File: stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift [hunk 3]
```diff
@@ -644,13 +647,16 @@ self.test("\(testNamePrefix).sorted/DispatchesThrough_withUnsafeMutableBufferPoi
...
+
   // This sort operation is not in-place.
   // The collection is copied into an array before sorting.
-  expectEqual(
-    0, lc.log._withUnsafeMutableBufferPointerIfSupported[type(of: lc)])
-  expectEqual(
-    0,
-    lc.log._withUnsafeMutableBufferPointerIfSupportedNonNilReturns[type(of: lc)])
+  expectEqual(0, actualWUMBPIF + actualWCMSIA)
+  expectEqual(0, actualWUMBPIFNonNil + actualWUMBPIFNonNil)
```
Comment (natecook1000) [inline stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift:659]:
```suggestion
  expectEqual(0, actualWUMBPIFNonNil + actualWCMSIAIFNonNil)
```

File: stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift [hunk 4]
```diff
@@ -965,11 +971,23 @@ self.test("\(testNamePrefix).partition/DispatchesThrough_withUnsafeMutableBuffer
...
+  let actualWCMSIA =  lc.log.withContiguousMutableStorageIfAvailable[type(of: lc)]
+
+  let actualWUMBPIFNonNil = lc.log._withUnsafeMutableBufferPointerIfSupportedNonNilReturns[type(of: lc)]
+  let actualWUMBPIFNonNil =  lc.log.withContiguousMutableStorageIfAvailableNonNilReturns[type(of: lc)]
+
+  expectEqual(1, actualWUMBPIF + actualWCMSIA)
   expectEqual(
     withUnsafeMutableBufferPointerIsSupported ? 1 : 0,
-    lc.log._withUnsafeMutableBufferPointerIfSupportedNonNilReturns[type(of: lc)])
+    actualWUMBPIFNonNil + actualWUMBPIFNonNil)
```
Comment (natecook1000) [inline stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift:983]:
```suggestion
    actualWUMBPIFNonNil + actualWCMSIAIFNonNil)
```

**Review by natecook1000 — APPROVED** [review]
This is great! 👏

File: stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift (hunk 1 above)
Comment (lorentey) [reply stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift:978]: Wow, these were really rough. Thanks for debugging all these; I should have paid more attention.

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
```

'WorkspaceTests.testArtifactDownloadHappyPath' started at 2021-02-24 06:35:06.861
/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-main/swiftpm/Tests/WorkspaceTests/WorkspaceTests.swift:4402: error: WorkspaceTests.testArtifactDownloadHappyPath : XCTAssertEqual failed: ("[<AbsolutePath:"/tmp/ws/.build/artifacts/A">, <AbsolutePath:"/tmp/ws/.build/artifacts/B">]") is not equal to ("[<AbsolutePath:"/tmp/ws/.build/artifacts/A">, <AbsolutePath:"/tmp/ws/.build/artifacts/A">, <AbsolutePath:"/tmp/ws/.build/artifacts/B">]") 
```

Hm.

**Comment by lorentey** [comment]
@swift-ci test linux platform

**Comment by lorentey** [comment]
@swift-ci test

**Pull request merged**